### ISSUE TITLE
On renewing the CG when working with multiple modules

### ISSUE THREAD

**Issue by saikrishnarallabandi** [clab/dynet#1049]
Hi,

When I use multiple modules within a single model, I notice that the training time increases significantly as the model is exposed to more training examples. I believe that there is something wrong in the way I am renewing the computational graph while combining the different modules but am unable to find it. Will appreciate help in debugging this. I am doing this:  For instance, if I have two different modules A and B such that A calls B, I am renewing the CG before calling A (for every example).  Here is an example module that uses 4 layer FFNNs as encoder and decoder networks in a VED ( Variational Encoder Decoder)

The loss function of the VED is as follows:
```
  def calculate_loss_basic(self, input, output):
       # Renew the computation graph
       #dy.renew_cg()  -> Not renewing the CG here but in the calling script that instantiates this.

       # Instantiate the params
       W_mean = dy.parameter(self.W_mean_p)
       V_mean = dy.parameter(self.V_mean_p)
       b_mean = dy.parameter(self.b_mean_p)
       W_var = dy.parameter(self.W_var_p)
       V_var = dy.parameter(self.V_var_p)
       b_var = dy.parameter(self.b_var_p)
       num_hidden = self.hidden_list[-1]
       
       enc_dnn = FeedForwardNeuralNet(self.model, [self.num_input, self.hidden_list, num_hidden, self.act])
       dec_dnn = FeedForwardNeuralNet(self.model, [num_hidden, self.hidden_list,self.num_output, self.act])

    
       # Get the output from encoder
       src_output = enc_dnn.predict(input)

       # Get the mean and log variance
       mu = self.mlp(src_output , W_mean, V_mean, b_mean)
       log_var = self.mlp(src_output , W_mean, V_mean, b_mean)       
    
       # Compute the KL Divergence loss
       kl_loss = -0.5 * dy.sum_elems( 1 + log_var - dy.pow(mu, dy.inputVector([2])) - dy.exp(log_var))

       # Reparameterize
       z = self.reparameterize(mu, log_var)
       
       return kl_loss, dec_dnn.calculate_loss(z , output)

```

The corresponding function in the FFNN is as follows:
```
  def predict(self, input):
    weight_matrix_array = []
    biases_array = []
    acts = []
    for (W,b, act) in zip(self.weight_matrix_array, self.biases_array, self.act):
         weight_matrix_array.append(dy.parameter(W))
         biases_array.append(dy.parameter(b))
         acts.append(act)
    g = acts[0]
    w = weight_matrix_array[0]
    b = biases_array[0]
    intermediate = g(w*input + b)
    activations = [intermediate]
    for (W,b, act) in zip(weight_matrix_array[1:], biases_array[1:], acts):
        pred =  act(W * activations[-1]  + b)
        activations.append(pred)
    return pred
```

The FFNN itself is declared as follows:

```
  def __init__(self, model, args):
    self.pc = model.add_subcollection()
    self.args = args
    self.num_input = int(args[0])
    self.num_output = int(args[2])
    self.hidden_list = args[1]
    self.act = args[3]
    self.model = model
    self.number_of_layers = len(self.hidden_list)
    num_hidden_1 = self.hidden_list[0]
    
    # Add first layer
    if debug == 1:
        print "Adding input to first hidden layer weights ", num_hidden_1, self.num_input
    self.W1 = self.pc.add_parameters((num_hidden_1, self.num_input))
    self.b1 = self.pc.add_parameters((num_hidden_1))
    
    # Add remaining layers
    self.weight_matrix_array = []
    self.biases_array = []
    self.weight_matrix_array.append(self.W1)
    self.biases_array.append(self.b1)
    for k in range(1, self.number_of_layers):
              if debug == 1: 
                   print "At ", k , " adding weights ", self.hidden_list[k], self.hidden_list[k-1]
              self.weight_matrix_array.append(self.model.add_parameters((self.hidden_list[k], self.hidden_list[k-1])))
              self.biases_array.append(self.model.add_parameters((self.hidden_list[k])))
    if debug == 1:
      print "Adding last hidden layer to output weights ", self.num_output, self.hidden_list[-1]
    self.weight_matrix_array.append(self.model.add_parameters((self.num_output, self.hidden_list[-1])))
    self.biases_array.append(self.model.add_parameters((self.num_output)))
    self.spec = (self.num_input, self.hidden_list, self.num_output, self.act)


```

The calling script is as follows:

```
     for (input_frame, output_frame) in zip(frames_handled_input, frames_handled_output):
       dy.renew_cg()  -> Renewing the CG here
       frame_count += 1
       count += 1
       KL_loss, RECONS_loss = ved.calculate_loss_basic(dy.inputTensor(input_frame), dy.inputTensor(output_frame))
       loss = dy.esum([KL_loss, RECONS_loss])
       train_loss += loss.value()
       kl_loss += KL_loss.value()
       recons_loss += RECONS_loss.value()
       loss.backward()
       if frame_count % update_params == 1:
            trainer.update() 
       if frame_count % 10000 == 1:
               print "Train Loss after epoch " +  str(epoch) + " and frames " + str(frame_count) + " : " +  str(float(train_loss/frame_count)) + " in files ", K , " of " , len(train_data)
               print "Recons_loss: ", float(recons_loss/frame_count)
               print "KL Loss: ", float(kl_loss/frame_count)
               end_time = time.time()
               print "Time: ", end_time - start_time

```

I have timed the module and here is the log file. It lists the time taken ( along with different losses) after processing every 10k audio frames. 

```
 Epoch  0
Train Loss after epoch 0 and frames 1 : 10.5394802094 in files  1  of  69
Recons_loss:  7.48326063156
KL Loss:  3.05621910095
Time:  0.0417790412903
Train Loss after epoch 0 and frames 10001 : 7.54742771572 in files  18  of  69
Recons_loss:  7.28733018536
KL Loss:  0.26009752844
Time:  71.477314949
Train Loss after epoch 0 and frames 20001 : 7.45011207987 in files  38  of  69
Recons_loss:  7.31722625267
KL Loss:  0.13288582683
Time:  267.409034967
Train Loss after epoch 0 and frames 30001 : 7.40307721421 in files  56  of  69
Recons_loss:  7.31291262832
KL Loss:  0.0901645860181
Time:  599.622094154
Train Loss after epoch 0 : 7.39242378033
Recons_loss:  7.319648312
KL Loss:  0.0727754690255


```

As can be seen, the time taken for each 10k frames keeps growing ( almost doubling) where as it should probably be increasing with a factor.

**Comment by saikrishnarallabandi** [comment]
Got it. The code I was using was trying to instantiate a new version of FFNN every time.  Making it a class variable and doing this resolved it. 

```


 enc_dnn = self.enc_dnn
 dec_dnn = self.dec_dnn


```
